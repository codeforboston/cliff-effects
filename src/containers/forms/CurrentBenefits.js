import { connect } from 'react-redux';

import { CurrentBenefitsStep } from '../../forms/CurrentBenefits';
import { setHasBenefit } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),

    USState: state.getIn([
      'geography',
      'state',
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setHasBenefit({ benefit, value }) {
      dispatch(
        setHasBenefit({
          time: 'current',
          benefit,
          value,
        })
      );
    },
  };
};

const CurrentBenefitsContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentBenefitsStep);

export { CurrentBenefitsContainer as CurrentBenefitsStep };
