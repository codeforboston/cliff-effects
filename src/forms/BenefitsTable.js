// REACT COMPONENTS
import React from 'react';
import {
  Table
} from 'semantic-ui-react';

// CUSTOM COMPONENTS
// Both the table and graph should just be added to a results page, but
// this will do for now

// BENEFIT LOGIC
import { getSNAPBenefits } from '../programs/federal/snap';
import { getHousingBenefit } from '../programs/state/massachusetts/housing';

// OBJECT MANIPULATION
import { cloneDeep } from 'lodash';


const getSignSymbol = function ( num ) {
  if ( num > 0 ) { return '+'; }
  else if ( num < 0 ) { return '-'; }
  else { return ''; }
};  // End getSignSymbol()


const BenefitsTable = function ( props ) {

  var currentClient = cloneDeep( props.client ),
      futureClient  = cloneDeep( currentClient );
  // Hack because everything's using future value instead
  // of just using current value.
  currentClient.future.earned = currentClient.current.earned;




  var SNAPBenefitCurrent  = Math.round( getSNAPBenefits( currentClient ).benefitValue * 12 ),
      SNAPBenefitFuture   = Math.round( getSNAPBenefits( futureClient ).benefitValue * 12 ),
      SNAPDiff            = SNAPBenefitFuture - SNAPBenefitCurrent,
      sec8BenefitCurrent  = Math.round( getHousingBenefit( currentClient ).benefitValue * 12 ),
      sec8BenefitFuture   = Math.round( getHousingBenefit( futureClient ).benefitValue * 12 ),
      sec8Diff            = sec8BenefitFuture - sec8BenefitCurrent,
      totalBenefitCurrent = SNAPBenefitCurrent + sec8BenefitCurrent,
      totalBenefitFuture  = SNAPBenefitFuture + sec8BenefitFuture,
      totalDiff           = SNAPDiff + sec8Diff,
      incomeCurrent       = Math.round( currentClient.current.earned * 12 ),
      incomeFuture        = Math.round( futureClient.future.earned * 12 ),
      incomeDiff          = incomeFuture - incomeCurrent,
      netCurrent          = totalBenefitCurrent + incomeCurrent,
      netFuture           = totalBenefitFuture + incomeFuture,
      netDiff             = totalDiff + incomeDiff;


const   columnHeaderStyle = {
                            background: 'rgba(0, 181, 173, 1)',
                            color: 'white',
                            fontSize: '1.3em',
                            fontWeight: 900,
                            textAlign: 'center',
                            borderRadius: 'inherit',
                            letterSpacing: '0.02em',
                            }
      , totalsRowStyle    = {
                            borderTop: "2px solid rgba(0, 181, 173, 1)",
                            fontWeight: 700,
                            fontSize: '1.1em',
                            padingTop: '0.25em'
                              }
      , rowHeaderStyle    = {
                    fontSize: '1.1em',
                    fontWeight: 700,
                    textAlign: 'left'
                    }
      , totalsRowHeaderStyle = {
                    fontSize: '1.2em',
                    fontWeight: 700,
                    textAlign: 'left',
                    borderTop: "2px solid rgba(0, 181, 173, 1)",
                    padingTop: '0.25em'


      };






const SNAPBenefitRow = function( props ){
  if(!currentClient.current.hasSnap ) return (null)
    return (
      <Table.Row>
        <Table.Cell style={rowHeaderStyle}>SNAP</Table.Cell>
        <Table.Cell textAlign='right'>${SNAPBenefitCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right'>${SNAPBenefitFuture} / year</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(SNAPDiff) } ${Math.abs(SNAPDiff)} / year</Table.Cell>
      </Table.Row>
    )
  };

const Sec8BenefitRow  = function( props ){  
  if(!currentClient.current.hasHousing) return (null)
    return (
      <Table.Row>
        <Table.Cell style={rowHeaderStyle}>Section 8 Housing</Table.Cell>
        <Table.Cell textAlign='right'>${sec8BenefitCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right'>${sec8BenefitFuture} / year</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(sec8Diff) } ${Math.abs(sec8Diff)} / year</Table.Cell>
      </Table.Row>
    ) 
};

const TotalBenefitsRow = function( props ){
  if(!currentClient.current.hasSnap || !currentClient.current.hasHousing) return (null)
    return(
      <Table.Row>
        <Table.Cell textAlign='right' width={3} style={totalsRowHeaderStyle}>Total Benefits</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${totalBenefitCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${totalBenefitFuture} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>{ getSignSymbol(totalDiff) } ${Math.abs(totalDiff)} / year</Table.Cell>
      </Table.Row>
    )
};

const IncomeRow = function ( props ) { 
    return (
      <Table.Row>
        <Table.Cell style={rowHeaderStyle}>Income</Table.Cell>
        <Table.Cell textAlign='right'>${incomeCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right'>${incomeFuture} / year</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(incomeDiff) } ${Math.abs(incomeDiff)} / year</Table.Cell>
      </Table.Row>
    )
};

const TotalsRow = function ( props ) {
    return (
      <Table.Row style={{border: 'none'}}>
        <Table.Cell textAlign='right' width={3} style={totalsRowHeaderStyle}>Net Total</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${netCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${netFuture} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>{ getSignSymbol(netDiff) } ${Math.abs(netDiff)} / year</Table.Cell>
      </Table.Row>
    )
  };


  return (
    <wrapper>
      <Table celled>
       <Table.Header>
          <Table.Row >
            <Table.Cell style={columnHeaderStyle} width={3}>Benefit</Table.Cell>
            <Table.Cell style={columnHeaderStyle} width={3}>Current Benefits</Table.Cell>
            <Table.Cell style={columnHeaderStyle} width={3}>New Estimate</Table.Cell>
            <Table.Cell style={columnHeaderStyle} width={3}>Difference</Table.Cell>
          </Table.Row>
      </Table.Header>
        <Table.Body>
          <SNAPBenefitRow client={currentClient} />
          <Sec8BenefitRow client={currentClient} />
          <TotalBenefitsRow client={currentClient} />
          <IncomeRow />
          <TotalsRow />
        </Table.Body>
      </Table>
    </wrapper>
  )

};  // End BenefitsTable(<>)


export {
	BenefitsTable
};
