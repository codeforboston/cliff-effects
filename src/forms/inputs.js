// REACT COMPONENTS
import React, { Component } from 'react';
import {
  Form,
  Button,
  Radio,
} from 'semantic-ui-react';


/** A big button. Right now, just used in the bottom row of
*     form sections
*
* @function
* @param {object} props
* @property {buttonProps} props.children - React children
* @property {buttonProps} props.buttonProps - Props to
*     override any default props that need to be overridden
*
* @returns Component
*/
const BigButton = function ({ children, ...buttonProps }) {

  const overriddenDefaults = {
    type:      `button`,
    color:     `teal`,
    size:      `large`,
    className: `big-button`,
    ...buttonProps,
  };

  return (
    <Button { ...overriddenDefaults }>{ children }</Button>
  );
};  // End <BigButton>


/** Yes/no toggleable radio button group with a label
 *
 * @function
 * @param {object} props
 * @property {string} props.labelText
 * @property {string} props.name - Key for radio-group. Must
 *     be unique from all other radio names on the page.
 * @property {bool} props.value - `true` if 'yes' is selected
 *     `false` if 'no' is selected. Change will be sent out.
 * @property {function} props.updateClientValue - is given event and adjusted
 *     input element props object. Adjustment is to make sure
 *     the property `value` is under control since there are
 *     issues further up the line.
 */
class ControlledRadioYesNo extends Component {
  constructor(props){
    super(props);
    this.state = {};
  }

  handleChange(e,inputProps){
    var obj = {
      ...inputProps,
      value: inputProps.label === `Yes`,
    };

    this.props.updateClientValue(e, obj);
  }


  render(){

    return (
      <div className="radio-yes-no">

        <Form.Field>
          <Radio
            label='Yes'
            name={ this.props.name }
            value='Yes'
            checked={ this.props.value === true }
            onChange={ this.handleChange.bind(this) } />
        </Form.Field>
        <Form.Field >
          <Radio
            label='No'
            name={ this.props.name }
            value='No'
            checked={ this.props.value === false }
            onChange={ this.handleChange.bind(this) } />
        </Form.Field>
        <Form.Field >
          <b>{this.props.labelText}</b>

        </Form.Field>

      </div>
    );
  }
};  // End <ControlledRadioYesNo>


/**
 * @todo description
 * @todo Write callback descriptions for function params: http://usejsdoc.org/tags-callback.html
 *
 * @param {Object} props
 * @param {number | string} props.value - Valid client value
 * @param {string} props.name - For HTML name property
 * @param {string} props.className - HTML class names
 * @param {*} [props.otherData] - Sent back to `store()`
 * @param {function} props.format - Given `value`. Must return what you want shown in the number field.
 * @param {function} props.validate - Given `value`. Must return boolean.
 * @param {function} props.store - Given an event, `value`, [`otherData`]
 */
class ManagedNumberField extends Component {
  constructor (props) {
    super(props);
    var { format, value } = props;
    this.state = { valid: true, focused: false, focusedVal: format(value) };
  }  // End constructor()

  //change form to blank string after click, before input
  handleFocus = (evnt) => {
    // This makes sure that only zeroes and blanks get reset
    var { format, value } = this.props;
    if (!Number.parseFloat(evnt.target.value)) {
      this.setState({ focused: true, focusedVal: '' });
    } else {
      this.setState({ focused: true, focusedVal: format(value) });
    }
  };

  handleBlur = (evnt) => {
    this.setState({ focused: false, valid: true });
  };

  handleChange = (evnt, inputProps) => {
    var { displayValidator, storeValidator, store, otherData } = this.props;
    var focusedVal = inputProps.value;

    // If doesn't pass display validator, don't store and don't change focusedVal
    if (!displayValidator(inputProps.value)) {
      return;
    }

    if (focusedVal.length === 0) {
      // If field contains an empty string, set value to be 0 (visible on blur)
      inputProps.value = '0';
    }
    var valid = storeValidator(inputProps.value);

    if (valid) {
      store(evnt, inputProps, otherData);
    }
    this.setState({ focusedVal: focusedVal, valid: valid });
  };  // End handleChange()

  render() {
    var { valid, focused, focusedVal }      = this.state;
    var { value, name, className, format }  = this.props;

    // Format correctly when neighbors are updated, if needed
    if (!focused) {
      value = format(value);
    } else {
      value = focusedVal;
    }

    /** @todo Different class for something 'future' that has a current value that isn't 0 */
    return (
      <Form.Input
        error     = { !valid }
        value     = { value }
        name      = { name }
        className = { className + ` output-number` }
        onChange  = { this.handleChange }
        onFocus   = { this.handleFocus }
        onBlur    = { this.handleBlur } />
    );
  }  // End render()

};  // End <ManagedNumberField>


export {
  BigButton,
  ControlledRadioYesNo,
  ManagedNumberField,
};
