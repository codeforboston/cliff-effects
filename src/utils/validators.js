/**
 * Set of validating functions for making sure
 * value are what they need to be. Just for
 * numbers right now.
 */


/** Returns true if a string represents a positve number (integer or float) */
// Should this only be valid if it has <= 2 decimal places?
const isPositiveNumber = function (str) {
  return str !== '' && !/[^0-9.]|\..*\./.test(str);
};


/** Returns true if a string represents a positive integer */
/** @todo Change name to 'isWholeNumber'. */
const isPositiveWholeNumber = function (str) {
  return str !== '' && /^[0-9]*$/.test(str);
};


export {
  isPositiveNumber,
  isPositiveWholeNumber,
};
