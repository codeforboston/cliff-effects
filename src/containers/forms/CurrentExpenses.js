import { connect } from 'react-redux';

import { CurrentExpensesStep } from '../../forms/CurrentExpenses';

import { setCashValue, setHousingType, setPaysUtility, setGetsFuelAssistance } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setExpenseValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },

    setHousingType({ housingType }) {
      dispatch(
        setHousingType({
          time: 'current',
          housingType,
        })
      );
    },

    setPaysUtility({ utility, paysUtility }) {
      dispatch(
        setPaysUtility({
          time: 'current',
          utility,
          paysUtility,
        })
      );
    },

    setGetsFuelAssistance({ getsAssistance }) {
      dispatch(
        setGetsFuelAssistance({
          time: 'current',
          getsAssistance,
        })
      );
    },
  };
};

const CurrentExpensesStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentExpensesStep);

export { CurrentExpensesStepContainer as CurrentExpensesStep };
