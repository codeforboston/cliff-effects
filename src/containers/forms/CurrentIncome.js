import { connect } from 'react-redux';

import { CurrentIncomeStep } from '../../forms/CurrentIncome';

import { setCashValue } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setIncomeValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },
  };
};

const CurrentIncomeStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentIncomeStep);

export { CurrentIncomeStepContainer as CurrentIncomeStep };
