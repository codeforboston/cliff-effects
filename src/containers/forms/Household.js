import { connect } from 'react-redux';

import { HouseholdStep } from '../../forms/Household';
import {
  addMember,
  removeMember,
  setMemberIsDisabled,
  setMemberRole,
  setMemberAge,
} from '../../actions';

const time = 'current';

const mapStateToProps = (state) => {
  return {
    household: state.getIn([
      'client',
      'current',
      'household', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setMemberAge({ index, age }) {
      dispatch(
        setMemberAge({
          time,
          index,
          age,
        })
      );
    },

    setMemberIsDisabled({ index, isDisabled }) {
      dispatch(
        setMemberIsDisabled({
          time,
          index,
          isDisabled,
        })
      );
    },

    setMemberRole({ index, role }) {
      dispatch(
        setMemberRole({
          time,
          index,
          role,
        })
      );
    },

    removeMember({ index }) {
      dispatch(
        removeMember({
          time,
          index,
        })
      );
    },

    addMember({ member }) {
      dispatch(
        addMember({
          time,
          member,
        })
      );
    },
  };
};

const HouseholdStepContainer = connect(mapStateToProps, mapDispatchToProps)(HouseholdStep);

export { HouseholdStepContainer as HouseholdStep };
