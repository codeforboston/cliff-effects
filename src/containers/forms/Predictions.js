import { connect } from 'react-redux';

import { PredictionsStep } from '../../forms/Predictions';

import { setCashValue } from '../../actions';

const mapStateToProps = (state) => {
  return { client: state.get('client') };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setPredictionValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'future',
          name,
          value,
        })
      );
    },
  };
};

const PredictionsStepContainer = connect(mapStateToProps, mapDispatchToProps)(PredictionsStep);

export { PredictionsStepContainer as PredictionsStep };
