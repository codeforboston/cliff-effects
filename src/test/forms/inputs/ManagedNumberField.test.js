import React from 'react';
import { mount } from 'enzyme';

import { ManagedNumberField } from '../../../forms/inputs';
import { isNonNegNumber, hasOnlyNonNegNumberChars } from '../../../utils/validators';


let getMNF = function (value, store, validator, displayValidator) {
  return (
    <ManagedNumberField
      value            = { value }
      store            = { store }
      storeValidator   = { validator }
      displayValidator = { displayValidator || validator }
      format           = { () => {} }
      name             = { `MNFTest` }
      otherData        = {{ interval: `monthly` }} />
  );
};


test('ManagedNumberField should render', () => {
  expect(() => {
    mount(getMNF(0));
  }).not.toThrow();
});

test('should set focused state to true on focus if current value of input is positive number', () => {
  const userInput = 10,
        wrapper   = mount(getMNF(userInput));
  // Really want this to be checked using an actual click, but apparently that's impossible :/
  wrapper.find('input').simulate('focus', { target: { value: userInput }});
  expect(wrapper.state('focused')).toBe(true);
});

// @todo Figure out why value of 0 creates warning
// test('should set focused state to true on focus if current value of input is 0', () => {
//   const userInput = 0,
//         wrapper   = mount(getMNF(userInput));
//   // The following line creates a warning
//   // 'A component is changing an uncontrolled input of type text to be controlled.'
//   wrapper.find('input').simulate('focus', { target: { value: userInput }});
//   expect(wrapper.state('focused')).toBe(true);
// });

// @todo Figure out why value of 0 creates warning
// test('should blank focusedVal state before focus if current value of input is 0', () => {
//   const userInput = 0,
//         wrapper   = mount(getMNF(userInput));
//   // The following line creates a warning
//   // 'A component is changing an uncontrolled input of type text to be controlled.'
//   wrapper.find('input').simulate('focus', { target: { value: userInput }});
//   expect(wrapper.state('focusedVal')).toBe('');
// });

test('should set focused state to false on blur', () => {
  const userInput = 10,
        wrapper   = mount(getMNF(userInput));
  wrapper.setState({ focused: true });
  wrapper.find('input').simulate('blur', { target: { value: userInput }});
  expect(wrapper.state('focused')).toBe(false);
});

test('should change local and app state correctly when user inputs positive number', () => {
  const mockValidator = jest.fn();
  const mockStore = jest.fn();

  // Define the test case here
  const userInput = 10;
  mockValidator.mockReturnValue(true);

  const wrapper = mount(getMNF(0, mockStore, mockValidator));
  wrapper.find('input').simulate('change', { target: { value: userInput }});

  // sending the proper call to change the application's state
  expect(mockStore.mock.calls[ 0 ][ 1 ].value).toEqual(userInput);

  // changes to this.state's focusedVal and valid variables
  expect(wrapper.state('focusedVal')).toBe(userInput);
  expect(wrapper.state('valid')).toBe(true);
});

test('should change local and app state correctly to 0 when user inputs empty string', () => {
  const mockValidator = jest.fn();
  const mockStore = jest.fn();

  // Define the test case here
  const userInput = '';
  mockValidator.mockReturnValue(true);

  const wrapper = mount(getMNF(10, mockStore, mockValidator));
  wrapper.find('input').simulate('change', { target: { value: userInput }});

  // sending the proper call to change the application's state
  // Since value is empty string, store should be given 0 as value
  expect(mockStore.mock.calls[ 0 ][ 1 ].value).toEqual(`0`);

  // changes to this.state's focusedVal and valid variables
  expect(wrapper.state('focusedVal')).toBe(userInput);
  expect(wrapper.state('valid')).toBe(true);
});

test('should change local and app state correctly to last valid value when user inputs negative number', () => {
  const mockStore = jest.fn();

  let mockDisplayValidator = jest.fn();
  mockDisplayValidator.mockImplementation(hasOnlyNonNegNumberChars);

  let mockValidator = jest.fn();
  mockValidator.mockImplementation(isNonNegNumber);

  // Define the test case here
  const userInput1 = 6;
  const userInput2 = -6;

  const wrapper = mount(getMNF(0, mockStore, mockValidator, mockDisplayValidator));
  wrapper.find('input').simulate('change', { target: { value: userInput1 }});
  wrapper.find('input').simulate('change', { target: { value: userInput2 }});

  // expect that the store has not been called
  expect(mockStore.mock.calls).toHaveLength(1);

  // changes to this.state's focusedVal and valid variables
  expect(wrapper.state('focusedVal')).toBe(userInput1);
  expect(wrapper.state('valid')).toBe(true);
});

test('should set FormInput `state.error` value correctly depending on `state.valid` value', () => {
  const wrapper = mount(getMNF(0));
  // Set valid to false -- FormInput error should be true
  wrapper.setState({ valid: false });
  expect(wrapper.find('FormInput').props().error).toBe(true);
  // Set valid back to true -- FormInput error should be false
  wrapper.setState({ valid: true });
  expect(wrapper.find('FormInput').props().error).toBe(false);
});
