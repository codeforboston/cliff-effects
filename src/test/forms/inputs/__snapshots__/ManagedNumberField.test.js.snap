// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManagedNumberField should match snapshot 1`] = `
<FormInput
  as={[Function]}
  className="undefined output-number"
  control={[Function]}
  error={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
/>
`;
